Add Ledger, Balance and Fees endpoints

nvcruzhe opened the pull request.

nvcruzhe committed:
Add balance and fees endpoints

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Add balance and fees endpoints

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Merge branch 'nvcruzheApiV3Local' of https://github.com/bitsoex/bitso-java into nvcruzheApiV3Local

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

Conflicts:
	src/main/java/com/bitso/Bitso.java
	src/main/java/com/bitso/exchange/Balance.java
	src/main/java/com/bitso/exchange/Fee.java

nvcruzhe committed:
Add Ledger endpoint

Add Ledger endpoint, with BitsoLedger and Operation classes.

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Add Ledger endpoint

Add Ledger endpoint, with BitsoLedger and Operation classes.

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

himerzi commented on src/main/java/com/bitso/BitsoLedger.java:
```diff
@@ -0,0 +1,30 @@
+    private Operation[] retrieveOperations(JSONArray array){
+        int totalElements = array.length();
+        Operation[] operations = new Operation[totalElements];
+        for(int i=0; i<totalElements; i++){
```
Porque no usar un "for each"?

himerzi commented on src/test/java/com/bitso/BitsoTest.java:
```diff
@@ -56,6 +56,30 @@ public void testUserAccountStatus() {
+        String[] operations = {"trades", "fees", "fundings", "withdrawals"};
+        BitsoLedger fullLedger = bitso.getUserLedger();
+        assertEquals(true, nullCheck(fullLedger, BitsoLedger.class));
+        for(int i=0; i<operations.length; i++){
```
Porque no usar un "for each"?

himerzi commented on src/main/java/com/bitso/helpers/Helpers.java:
```diff
@@ -100,4 +103,21 @@ public static BigDecimal getBD(JSONObject o, String key) {
         return null;
     }
+
+    public static Date getDate(String s){
```
Aqui hay que usar un ZonedDateTime en vez de Date, es la clase que aqui usamos para representar fechas. Si te fijas el "received string" tiene informacion de zona horario que se pierde al utilizar Date.

nvcruzhe committed:
Merge branch 'nvcruzheApiV3Local' of https://github.com/bitsoex/bitso-java into nvcruzheApiV3Local

Conflicts:
	src/main/java/com/bitso/exchange/Operation.java
	src/main/java/com/bitso/helpers/Helpers.java
	src/test/java/com/bitso/BitsoTest.java

nvcruzhe committed:
Add withdrawal endpoint

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Add funding endpoints

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Add user trades endpoint

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

himerzi commented on src/main/java/com/bitso/Bitso.java:
```diff
@@ -161,16 +161,25 @@ public BitsoFee getUserFees(){
+            return null;
         }
+        return new BitsoLedger(o);
+    }
```
Para minimizar repeticion de codigo, que opinas de lo siguiente?

```
public BitsoLedger getUserLedger(){
         return getUserLedger("");
}

```

vogelito replied:
Creo que siempre es mejor pasar `null` (and handle that way) en vez de mandar empty strings, no?

nvcruzhe replied:
valueOf(object) se usa para hacer el concat de strings, por lo que lo que se concatenaria  un "null" dando un resultado no esperado, sin embargo se puede modificar la funcion para validar que el parametro pasado sea diferente de null y de esa forma evitamos la duplicidad de codigo.

himerzi replied:
(habia comentado esto, pero aparecio abajo) El pedo es que concatenar null a un string literalmente concatena el string
"null". Si crees que es mejor pasar null hay que modificar la función para
que cheque si es null y concatene "". What Victor said!

himerzi commented on src/main/java/com/bitso/Bitso.java:
```diff
@@ -208,12 +217,32 @@ public BitsoFunding getUserFundings(String... fundingsIds){
+            for (String string : tradesIds) {
+                request += string + "-";
+            }
+            request = request.substring(0, request.length() - 1);
```
Porque andas haciendo `request = request.substring(0, request.length() - 1);`?

himerzi replied:
Aaaa...ya entendi. Que tal en vez de tener esa linea, utilizar un for tradicional y detenerte una iteracion antes? Realmente performance aqui no importa, pero si se me hace mas claro hacerlo de esa manera.

nvcruzhe replied:
previous for leaves request string with a "-" at the end, substring funtion removes that character in order to get the correct string.

himerzi commented:
El pedo es que concatenar null a un string literalmente concatena el string
"null". Si crees que es mejor pasar null hay que modificar la función para
que cheque si es null y concatene "".
On Tue, Jan 17, 2017 at 6:23 PM Daniel Vogel <notifications@github.com>
wrote:

> *@vogelito* commented on this pull request.
>
>
>
>
> ------------------------------
>
>
>
>
> In src/main/java/com/bitso/Bitso.java
> <https://github.com/bitsoex/bitso-java/pull/10>:
>
>
> >          }
>
> +        return new BitsoLedger(o);
>
> +    }
>
>
>
> Creo que siempre es mejor pasar null (and handle that way) en vez de
> mandar empty strings, no?
>
>
>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitsoex/bitso-java/pull/10>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAy-EDgGk5itgKU-mMPFVj2EjwAnX4zBks5rTVt5gaJpZM4Ljado>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

nvcruzhe committed:
Add Order endpoints

Add Order endpoints OpenOrders, LookupOrders, Cancel Order, Place Order
Add support functions to helpers.

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

himerzi commented on src/main/java/com/bitso/Bitso.java:
```diff
@@ -176,6 +176,44 @@ public BitsoLedger getUserLedger(String... specificOperation){
+        String request = "/api/v3/withdrawals";
+        if(withdrawalsIds.length > 0){
+            request += "/";
+            for (String string : withdrawalsIds) {
```
Que opinas si en vez de ponerle de mas y despues tener que cortar el string, haces:
`for (int i = 0; i < withdrawalsIds.length - 1; i++)`
y luego afuera del for:
`request += withdrawalsIds[withdrawalsIds.length]`
Asi el String request queda a la medida y no tenemos que hacer operaciones extras.

himerzi replied:
Lo podriamos hacer asi en todos los codigos con este patron.

himerzi commented on src/main/java/com/bitso/exceptions/NotExpectedValue.java:
```diff
@@ -0,0 +1,7 @@
+package com.bitso.exceptions;
+
+public class NotExpectedValue extends RuntimeException {
```
Nice!

himerzi commented on src/main/java/com/bitso/exchange/Order.java:
```diff
@@ -0,0 +1,121 @@
+            case "eth_mxn":
+                return BitsoBook.ETH_MXN;
+            default:
+                String exceptionMessage = book + "is not a supported value";
```
..."is not a supported *book*"?

himerzi commented on src/main/java/com/bitso/exchange/Order.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    private Order.SIDE getSide(String side) {
+        if (side.compareTo("buy") == 0) {
```
Porque no usar .equals() ?

himerzi commented on src/main/java/com/bitso/exchange/Order.java:
```diff
@@ -0,0 +1,121 @@
+            case "cancelled":
+                return Order.STATUS.CANCELLED;
+            default:
+                String exceptionMessage = status + "is not a supported value";
```
...."is not a supported order state" para ser mas explicitos

himerzi commented on src/main/java/com/bitso/exchange/Order.java:
```diff
@@ -0,0 +1,121 @@
+            case "market":
+                return Order.TYPE.MARKET;
+            default:
+                String exceptionMessage = type + "is not a supported value";
```
...."is not a supported order type" para ser mas explicitos

himerzi commented on src/main/java/com/bitso/exchange/Order.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    private ZonedDateTime verifyUpdateDate(Object o) {
+        if (o != null) {
```
Porque el null check? Hay instancias donde updated_at es nulo? Si es el caso, sugiero no usar o.get() en el constructor de Order, ya que puede tirar una exception si la llave no esta presente. Podrias usar o.getString() en el constructor, y aqui tambien cambiar el parametro de object a string. Aqui podrias checar si el string es nulo o vacio.

En general no me encanta la idea de perder el sistema de tipaje de Java a favor de usar Object, ya sabes?

himerzi commented on src/main/java/com/bitso/helpers/Helpers.java:
```diff
@@ -118,8 +118,14 @@ public static Integer getInteger(JSONObject o, String key){
 
-    public static ZonedDateTime getZonedDatetime(String s){
-        return ZonedDateTime.parse(s, dateTimeFormatter);
+    public static ZonedDateTime getZonedDatetime(JSONObject o, String key){
```
Ah, ya vi que agregaste esto! Porque no usarlo donde estas usando `verifyUpdateDate()` y borrar esa funcion?

nvcruzhe committed:
Add missing endpoints

Add missing endpoints: Palce order, Funding destination, bitcoin
and ether withdrawals, Get Code Banks, Debit Card Withdrawal.

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Complete endpoints and remove unused classes

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Complete endpoints and remove unused classes

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Merge branch 'nvcruzheApiV3Local' of https://github.com/bitsoex/bitso-java into nvcruzheApiV3Local

himerzi commented on src/test/java/com/bitso/BitsoTest.java:
```diff
@@ -42,124 +38,139 @@ public void testTicker() {
 
     // Private endpoints
-    //@Test
+    // @Test
```
Hay que dejar todos los tests sin comentarios.

mariorz commented on src/main/java/com/bitso/BitsoBank.java:
```diff
@@ -0,0 +1,20 @@
+
+import com.bitso.helpers.Helpers;
+
+public class BitsoBank {
```
Consider if it would be more convenient for the response of getBitsoBanks() to be a HashMap, so that when you use it calling debit_card_withdrawal endpoint you can directly access the code for your bank without requiring an iteration over every bank.

nvcruzhe committed:
Delete unused classes

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Last changes required

Cleaning the code, changing BitsoBank implementation, deleting
unused classes

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

nvcruzhe committed:
Last changes required

Cleaning the code, changing BitsoBank implementation, deleting
unused classes

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>

mariorz commented:
Hey Victor, I think this is looking pretty good! 
You should probably check if the README.md (documentation) is up to date or requires some changes.

nvcruzhe committed:
Merge branch 'nvcruzheApiV3Local' of https://github.com/bitsoex/bitso-java into nvcruzheApiV3Local

Conflicts:
	src/main/java/com/bitso/BitsoTrade.java

nvcruzhe committed:
Remove TODO comments

Missing functionality is already implemented and reviewed against
server requests.

Signed-off-by: Nelson Victor Cruz Hdez <victor.cruz.isc@gmail.com>
